### PR TITLE
Use "produce" and "consume" terminology throughout.

### PR DESCRIPTION
**Pull request by jmalloc**
I'm also addressing #60 as I go.

### CODE REVIEW

File: integration.go [hunk 1]
```diff
@@ -4,22 +4,37 @@ import "context"
...
 	//
-	// c provides access to the various configuration options, such as specifying
-	// which types of integration command messages are routed to this handler.
+	// The engine MUST call Configure() before calling HandleEvent(). It is
+	// RECOMMENDED that the engine only call Configure() once per handler.
 	Configure(c IntegrationConfigurer)
 
-	// HandleCommand handles an integration command message that has been routed to
-	// this handler.
+	// HandleEvent handles a command message.
```
Comment (koden-km) [inline integration.go]:
```suggestion
	// HandleCommand handles a command message.
```

File: projection.go [hunk 2]
```diff
@@ -61,18 +63,18 @@ type ProjectionConfigurer interface {
...
+	// ConsumesEventType configures the engine to route event messages of the
 	// same type as m to the handler.
 	//
 	// It MUST be called at least once within a call to Configure(). It MUST NOT
 	// be called more than once with an event message of the same type.
 	//
-	// Multiple handlers within a single application MAY receive event messages
-	// of the same type.
+	// Multiple handlers within an application MAY consume event messages of the
+	// same type.
```
Comment (danilvpetrov) [inline projection.go]: So far all the changes make sense to me. I just wonder if the statement above will end up in process iface descriptions. I assume the "produce" and "consume" terminology terminology did not make it to aggregates and processes yet.

File: integration.go [hunk 3]
```diff
@@ -32,28 +47,52 @@ type IntegrationMessageHandler interface {
...
+	// the same type as m to the handler.
+	//
+	// It MUST be called at least once within a call to Configure(). It MUST NOT
+	// be called more than once with a command message of the same type.
+	//
+	// A given command type MUST be routed to exactly one handler within an
+	// application.
+	//
+	// The "content" of m MUST NOT be used, inspected, or treated as meaningful
+	// in any way, only its runtime type information.
```
Comment (koden-km) [inline integration.go]: This last part feels a bit off.
> only its runtime type information.

Maybe end the sentence and reference "m" again? Something like:
> in any way. Only the runtime type information of "m" should be used.

File: aggregate.go [hunk 4]
```diff
@@ -11,47 +11,63 @@ package dogma
...
-	// handler.
+	// Configure produces a configuration for this handler by calling methods on
+	// the configurer c.
+	//
+	// The implementation MUST allow for multiple calls to Configure(). Each
+	// call SHOULD produce the same configuration.
 	//
-	// c provides access to the various configuration options, such as specifying
-	// which types of domain command messages are routed to this handler.
+	// The engine MUST call Configure() before calling HandleEvent(). It is
```
Comment (koden-km) [inline aggregate.go]: This should be `HandleCommand()`?
> before calling HandleEvent()

File: aggregate.go [hunk 5]
```diff
@@ -60,6 +76,12 @@ type AggregateMessageHandler interface {
 type AggregateRoot interface {
 	// ApplyEvent updates the aggregate instance to reflect the occurence of an
 	// event that was recorded against this instance.
+	//
+	// It MUST NOT be called with a message of any type that has not been
+	// configured for production by a prior call to Configure().
+	//
+	// It MUST accept all messages ofthe types  that have been configured for
```
Comment (koden-km) [inline aggregate.go]: Missing space `ofthe`, and added space `types  that`:
`// It MUST accept all messages ofthe types  that have been configured for`

Fix:
`// It MUST accept all messages of the types that have been configured for`

File: aggregate.go [hunk 6]
```diff
@@ -72,77 +94,96 @@ type AggregateRoot interface {
...
+	// the same type as m to the handler.
+	//
+	// It MUST be called at least once within a call to Configure(). It MUST NOT
+	// be called more than once with a command message of the same type.
+	//
+	// A given command type MUST be routed to exactly one handler within an
+	// application.
+	//
+	// The "content" of m MUST NOT be used, inspected, or treated as meaningful
+	// in any way, only its runtime type information.
```
Comment (koden-km) [inline aggregate.go]: If changing this type info part in the other review comment, then this should also change to match.

File: projection.go (hunk 2 above)
Comment (jmalloc) [reply projection.go]: I've done aggregates now, but not processes yet.

File: aggregate.go (hunk 5 above)
Comment (jmalloc) [reply aggregate.go]: My spacebar key is packing it in on this new laptop :(

Comment (koden-km) [reply aggregate.go]: You need a space bar button on the touchbar.

File: process.go [hunk 7]
```diff
@@ -123,91 +163,92 @@ type ProcessEventScope interface {
...
+	// It MUST NOT be called if the instance has not begun.
 	//
-	// The precise semantics of ending a process instance are implementation
-	// defined. The engine is not required to allow re-beginning a process
-	// instance that has been ended.
+	// The engine MUST discard any timeout messages associated with this
+	// instance.
+	//
+	// The engine MAY allow re-beginning a process instance that
+	// has ended. Callers SHOULD assume that such behavior is unavailable.
```
Comment (koden-km) [inline process.go]: This "has" could have fit on the previous line? Could try reflow this doc block again.

File: process.go [hunk 8]
```diff
@@ -20,60 +20,79 @@ import (
...
 	// Handling an event message involves inspecting the state of the target
 	// process instance to determine what command messages, if any, should be
 	// produced.
 	//
-	// s provides access to the operations available within the scope of handling
-	// m, such as beginning or ending the targeted instance, accessing its state,
-	// sending command messages or scheduling timeouts.
+	// The engine SHOULD provide "at-least-once" delivery guarantees to the
+	// handler. That is, the engine should call HandleEvent() with the same
+	// command message until a nil error is returned.
```
[the remark this answers is not in the document]
Comment (koden-km) [reply process.go]: It doesn't allow the suggestions thing during partial diff views or something.

**Comment by jmalloc** [comment]
I think I've addressed all of the mad pirate's concerns.

**Comment by koden-km** [comment]
Ok the diff looks good now. As discuss I will read over all the docs as plain files (non diffed) and see if it all makes sense before giving the final approval.

**Comment by jmalloc** [comment]
Cool, thanks.

**Comment by jmalloc** [comment]
If I type in here, and press enter, it makes Danil's laptop go "ba-ding". Listen carefully.

**Comment by danilvpetrov** [comment]
That's actually my phone 😄

**Comment by danilvpetrov** [comment]
... and it's silent now.

**Comment by koden-km** [comment]
Were you going to update the docs in `application.go` also? It's `Configure()` method doesn't have the strict language.

**Comment by koden-km** [comment]
https://github.com/dogmatiq/dogma/blob/consume-produce-terminology/integration.go#L17
The docs say `HandleEvent()` but that method doesn't exist on this interface. Should it be `HandleCommand()`?

**Comment by koden-km** [comment]
https://github.com/dogmatiq/dogma/blob/consume-produce-terminology/process.go#L37
This one is the opposite to the previous comment. Maybe those are supposed to be like that then.

**Comment by koden-km** [comment]
Those 3 comments above are the only things that stood out after re-reading all the code docs. They are the only things i'm unsure about before I mark approved.

**Comment by jmalloc** [comment]
> Were you going to update the docs in `application.go` also? It's `Configure()` method doesn't have the strict language.

I was just changing up the handler docs in this PR, since they had to be changed for the terminology change anyway.

**Comment by jmalloc** [comment]
You were right about the other two; they have been fixed.

**Pull request merged**